crmsh/utils.py

```diff
@@ -752,6 +830,32 @@ def mkdirp(directory, mode=0o777, parents=True, exist_ok=True):
     """
     Path(directory).mkdir(mode, parents, exist_ok)
 
+def mkdirs_owned(dirs, mode=0o777, uid=-1, gid=-1):
+    """
+    Create directory path, setting the mode and
+    ownership of the leaf directory to mode/uid/gid.
+    It won't fail if the directory already exist (exist_ok===true).
+    """
+    if not os.path.exists(dirs):
+        try:
+            if not os.path.exists(dirs):
+                os.makedirs(dirs, mode)
+        except OSError as err:
+            # If we failed under current user, repeat under root
+            cmd = "sudo mkdir {}".format(dirs)
+            cmd += " && sudo chmod {} {}".format(format(mode,'o'), dirs)
+            if gid != -1 and uid != -1:
+                cmd += " && sudo chown {}:{} {}".format(uid, gid, dirs)
+            elif gid == -1 and uid != -1:
+                cmd += " && sudo chown {} {}".format(uid, dirs)
+            elif gid != -1 and uid == -1:
+                cmd += " && sudo chgrp {} {}".format(gid, dirs)
+            rc, out, err = get_stdout_stderr(cmd, no_reg=True)
+            if rc != 0:
+                fatal("Failed to create {}: {}".format(' '.join(dirs), err))
+            return
```

Comment:
Rewriting every `os.*` function with shell scripts is complex and error-prone. Can we try to spawn a python process with sudo and call these function with it?

Reply:
Yes, this is possible. There are basically two things we can do, basically either

- run the whole crm process as root, and do the exceptions when it should not be root, i.e. when ssh'ing to another node.

or

- run everything as a non-root and do exceptions when the root is required.

I'm more inclined to get rid of the root as much as we can, (that's why I chose the second solution). We don't really need to rewrite all os functions. There are just a few (read/write file, chmod, chown) and they are already in this PR. Besides, we already heavily use bash in crmsh (also for reading, writing, chmod, chown) and it works as a charm.